account_payment_batch_oca/models/account_payment_line.py

```diff
@@ -0,0 +1,342 @@
+    )
+    communication_type = fields.Selection(
+        compute="_compute_payment_line",
+        store=True,
+        readonly=False,
+        precompute=True,
+        selection=[("free", "Free"), ("structured", "Structured")],
+        required=True,
+    )
+    payment_ids = fields.Many2many(
+        comodel_name="account.payment",
+        string="Payment transaction",
+        readonly=True,
+    )
+
+    _sql_constraints = [
+        (
+            "name_company_unique",
+            "unique(name, company_id)",
+            "A payment line already exists with this reference in the same company!",
+        )
+    ]
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            if vals.get("name", "New") == "New":
+                vals["name"] = (
+                    self.env["ir.sequence"].next_by_code("account.payment.line")
+                    or "New"
+                )
+        return super().create(vals_list)
+
+    @api.depends("amount_currency", "currency_id", "company_currency_id", "date")
+    def _compute_amount_company_currency(self):
+        for line in self:
+            if line.currency_id and line.company_currency_id:
+                line.amount_company_currency = line.currency_id._convert(
+                    line.amount_currency,
+                    line.company_currency_id,
+                    line.company_id,
+                    line.date or fields.Date.today(),
+                )
+            else:
+                line.amount_company_currency = 0
+
+    @api.model
+    def _lot_grouping_fields(self):
+        """This list of fields is used to compute the grouping hashcode for lots.
+        This method is inherited in account_payment_sepa_base to add several fields.
+        The fields in this list MUST also be present in _payment_grouping_fields()
+        """
+        return [
+            "date",
+            "currency_id",
+        ]
+
+    @api.model
+    def _payment_grouping_fields(self):
+        """This list of fields is used o compute the grouping hashcode."""
+        res = self._lot_grouping_fields()
+        res += [
+            "partner_id",
+            "partner_bank_id",
+            "communication_type",
+        ]
+        return res
+
+    def payment_line_hashcode(self):
+        self.ensure_one()
+        values = []
+        for field in self._payment_grouping_fields():
+            values.append(str(self[field]))
+        # Don't group the payment lines that are attached to the same supplier
+        # but to move lines with different accounts (very unlikely),
+        # for easier generation/comprehension of the transfer move
+        values.append(str(self.move_line_id.account_id or False))
+        # Don't group the payment lines that use a structured communication
+        # otherwise it would break the structured communication system !
+        if self.communication_type != "free":
+            values.append(str(self.id))
```

Comment:
```suggestion
        if self.communication_type != "free":
            values.append(str(self.id))
        else:
            values.append("")
```

So we have the same number of `-` in all grouping keys.

Reply:
Now that we use tuples, there is no more "-" in the grouping keys

Reply:
Yes but the tuples should always have the same size, so let's add None?